### PR TITLE
[v3] a light round of golf

### PR DESCRIPTION
**Pull request by leeoniya**
did a bit of perf optimization and legible golfing.

also added baidu bot (i see it all the time in my server logs)

### CODE REVIEW

File: src/yall.mjs [hunk 1]
```diff
@@ -63,7 +63,7 @@ export default function (options) {
 
   // If the current user agent is a known crawler, immediately load all media
   // for the elements yall is listening for and halt execution (good for SEO).
-  if (/(google|bing|yandex|duckduck)bot/i.test(navigator.userAgent)) {
+  if (/baidu|(?:google|bing|yandex|duckduck)bot/i.test(navigator.userAgent)) {
```
Comment (malchata) [inline src/yall.mjs]: Is this syntax widely supported? I can't recall if this is specific to ES6 or if has been around forever.

Comment (leeoniya) [reply src/yall.mjs]: yeah, forever i think. the fancy stuff is named capturing groups/backrefs, lookbehinds. i think positionl backreferences had some issues back in es3.

http://blog.stevenlevithan.com/archives/fixing-javascript-regexp

File: src/yall.mjs [hunk 2]
```diff
@@ -36,24 +40,26 @@ export default function (options) {
       element.load();
     }
 
+    const classList = element.classList;
```
Comment (malchata) [inline src/yall.mjs]: that's heckin' smart

File: src/yall.mjs [hunk 3]
```diff
@@ -3,27 +3,31 @@ export default function (options) {
     options = {};
   }
 
-  const intersectionObserverSupport = "IntersectionObserver" in window && "IntersectionObserverEntry" in window;
+  const doc = document;
```
Comment (malchata) [inline src/yall.mjs]: I like these. They're still readable, but the uglifier will do the same thing with them as my abbreviations did.

File: src/yall.mjs (hunk 2 above)
Comment (malchata) [inline src/yall.mjs]: That's heckin' smart.

File: src/yall.mjs [hunk 4]
```diff
@@ -36,24 +40,26 @@ export default function (options) {
...
+    if (classList.contains(lazyBackgroundClass)) {
+      classList.remove(lazyBackgroundClass);
+      classList.add(options.lazyBackgroundLoaded || "lazy-bg-loaded");
     }
   };
 
   // Added because there was a number of patterns like this peppered throughout
   // the code. This just flips necessary data- attrs on an element
   const yallFlipDataAttrs = element => {
-    ["srcset", "src", "poster"].forEach(dataAttr => {
```
Comment (malchata) [inline src/yall.mjs]: Hm. Last I did this, it added some bytes. But, this approach means we're not instantiating an array every time we flip data attributes. This is probably more sound.

**Comment by malchata** [comment]
Hey, @leeoniya. I _just_ pushed some code to this branch. Can you rebase your changes on top of this branch?

File: src/yall.mjs [hunk 5]
```diff
@@ -3,27 +3,31 @@ export default function (options) {
...
+
+  const intersectionObserverSupport = "IntersectionObserver" in win && "IntersectionObserverEntry" in win;
   const lazyClass = options.lazyClass || "lazy";
   const lazyBackgroundClass = options.lazyBackgroundClass || "lazy-bg";
   const idleLoadTimeout = "idleLoadTimeout" in options ? options.idleLoadTimeout : 100;
   const observeChanges = options.observeChanges || false;
   const selectorString = `img.${lazyClass},video.${lazyClass},iframe.${lazyClass},.${lazyBackgroundClass}`;
 
-  // This abstraction shaves off a few bytes (plus it's nifty).
-  const sliceCall = arr => [].slice.call(arr);
```
Comment (malchata) [inline src/yall.mjs]: NodeLists aren't iterable in IE11 using forEach :\ There may be some further opportunities here, though to see if maybe we can iterate over them with `for` instead. In some cases, I've found that going from `forEach` to `for` can lop off a few bytes.

**Comment by malchata** [comment]
Hey, @leeoniya.

Actually, I may want to sort of cherry pick some optimizations here and bring them in manually. I think the code golfing stuff is good, but since version 3 will be in active development for a while, I may not want to pull in everything just to keep things maintainable. Does that work?

Sorry if it seems like I'm jerking you around. I really appreciate your contributions to this project.

**Comment by malchata** [comment]
ack, closed prematurely. Sorry!

**Comment by leeoniya** [comment]
no problem. i'm not expecting get visible commits into this project, feel free to cherry pick and close this PR, i opened it cause it was the best way to show you the changes.

> I may not want to pull in everything just to keep things maintainable

i don't think any of these changes make it less maintainable, but do as you wish :)

**Comment by malchata** [comment]
I really appreciate your help and enthusiasm, @leeoniya. You've been really cool about everything. :)

**Comment by leeoniya** [comment]
i like micro-libs that are written efficiently and do one thing well [1] (unix philisophy). since i'm using yall, i want to make it better where i can.

[1] i made this a couple days ago: https://github.com/leeoniya/dropcss

File: src/yall.mjs (hunk 5 above)
Comment (malchata) [reply src/yall.mjs]: Wow, I totally did not read this right. I see what you did here and this makes a lot of sense. Nevermind that first comment. I'm totally bringing this in.

Comment (leeoniya) [reply src/yall.mjs]: yeah, it's still doing the same thing.

Comment (malchata) [reply src/yall.mjs]: heckin' smart!

Comment (leeoniya) [reply src/yall.mjs]: nah, i'm a pretty average golfer. you should see the crazy shit the Preact team has to do to shave every byte off.

also:

https://js1k.com/
https://www.dwitter.net/

Comment (malchata) [reply src/yall.mjs]: I'm a reasonably OK, but amateur golfer, but I've looked at the PRs for preact. It's all shit like "-4B" but it all adds up.

omg TIL about dwitter. This is so cool, thanks for sharing it!

**Pull request closed without merging**